File: Delivery System 2/tariff.cpp
#include "tariff.h"
#include "parcel.h"
#include "address.h"

// Ѕазовый тариф
Tariff::Tariff(const string& name, double basePrice)
	: m_name(name), m_basePrice(basePrice) {
}

// Ёкспресс тариф - дороже но быстрее
ExpressTariff::ExpressTariff(const string& name, double basePrice)
	: Tariff(name, basePrice) {
}

double ExpressTariff::calculateCost(const Parcel& parcel, const Address& from, const Address& to) {
	return m_basePrice + parcel.calculateVolume() * VOLUME_RATE + parcel.getWeight() * WEIGHT_RATE;
}

// Ёконом тариф - дешевле но медленнее
EconomyTariff::EconomyTariff(const string& name, double basePrice)
	: Tariff(name, basePrice) {
}

double EconomyTariff::calculateCost(const Parcel& parcel, const Address& from, const Address& to) {
	return m_basePrice + parcel.calculateVolume() * VOLUME_RATE + parcel.getWeight() * WEIGHT_RATE;
}

File: Delivery System 2 Java/src/Tariff.java
package src;

public abstract class Tariff {
    protected String name;
    protected double basePrice;

    public Tariff(String name, double basePrice) {
        this.name = name;
        this.basePrice = basePrice;
    }

    // Абстрактный метод расчета стоимости
    public abstract double calculateCost(Parcel parcel, Address from, Address to);

    // Геттеры
    public String getName() { return name; }
    public double getBasePrice() { return basePrice; }
}